pep-0612.rst

```diff
@@ -128,169 +156,346 @@ will defer the standardization of the semantics of those options to a later PEP.
 
 With those requirements met, we can now take advantage of the unique properties
-afforded to us by this set up: 
+afforded to us by this set up:
 
 
-* Inside the function, ``args`` has the type ``TParams.args``\ , not 
-  ``Tuple[TParams.args, ...]`` as would be with a normal annotation 
+* Inside the function, ``args`` has the type ``P.args``\ , not
+  ``Tuple[P.args, ...]`` as would be with a normal annotation
   (and likewise with the ``**kwargs``\ )
-* A function of type ``Callable[TParams, TReturn]`` can be called with 
-  ``(*args, **kwargs)`` if and only if ``args`` has the type ``TParams.args`` 
-  and ``kwargs`` has the type ``TParams.kwargs``\ , and that those types both 
-  originated from the same function declaration.
-* A function declared as 
-  ``def inner(*args: TParams.args, **kwargs: TParams.kwargs) -> X``
-  has type ``Callable[TParams, X]``.
+* A function of type ``Callable[P, R]`` can be called with ``(*args, **kwargs)``
+  if and only if ``args`` has the type ``P.args`` and ``kwargs`` has the type
+  ``P.kwargs``\ , and that those types both originated from the same function
+  declaration.
+* A function declared as ``def inner(*args: P.args, **kwargs: P.kwargs) -> X``
+  has type ``Callable[P, X]``.
 
 With these three properties, we now have the ability to fully type check
 parameter preserving decorators.
 
-One additional form that we want to support is functions that pass only a subset
-of their arguments on to another function. To avoid shadowing a named or keyword
-only argument in the ``ParameterSpecification`` we require that the additional
-arguments be anonymous arguments that precede the ``*args`` and ``*kwargs``
+.. code-block::
+
+   def decorator(f: Callable[P, int]) -> Callable[P, None]:
+
+     def foo(*args: P.args, **kwargs: P.kwargs) -> None:
+
+       f(*args, **kwargs)    # Accepted, should resolve to int
+
+       f(*kwargs, **args)    # Rejected
+
+       f(1, *args, **kwargs) # Rejected
+
+     return foo              # Accepted
+
+To extend this to include ``Concatenate``, we declare the following properties:
+
+* A function of type ``Callable[Concatenate[A, B, P], R]`` can be
+  called with ``(a, b, *args, **kwargs)`` when ``args`` and ``kwargs`` are the
+  respective components of ``P``, ``a`` is of type ``A`` and ``b`` is of
+  type ``B``.
+* A function declared as
+  ``def inner(a: A, b: B, *args: P.args, **kwargs: P.kwargs) -> R``
+  has type ``Callable[Concatenate[A, B, P], R]``.
 
 .. code-block::
 
-   def call_n_times(
-       __f: Callable[TParams, None], 
-       __n: int, 
-       *args: TParams.args, 
-       **kwargs: TParams.kwargs,
-   ) -> None:
-       for x in range(__n);
-           __f(*args, **kwargs)
+   def add(f: Callable[P, int]) -> Callable[Concatenate[str, P], None]:
+
+     def foo(s: str, *args: P.args, **kwargs: P.kwargs) -> None:  # Accepted
+       pass
+
+     def bar(*args: P.args, s: str, **kwargs: P.kwargs) -> None:  # Rejected
+       pass
```

Comment:
nit: add an bullet point before that explicitly point out this restriction?

Reply:
I discuss it down on line 480

Reply:
I saw it but thought it was too late. It might not be worthwhile to explain why you introduce this restriction here but I'd expect the bullet points before to mention the existence of the restriction. 

I was expecting the code example to reinforce what was mentioned before in the text. But it turns out that it actually introduces the said restriction -- the text didn't mention it at all. Was a bit surprised by it.